enhancements/machine-api/out-of-tree-provider-support.md

```diff
@@ -0,0 +1,477 @@
+- vSphere: https://github.com/kubernetes/cloud-provider-vsphere
+vSphere introduced an additional service to expose internal API server: https://github.com/kubernetes/cloud-provider-vsphere/blob/master/manifests/controller-manager/vsphere-cloud-controller-manager-ds.yaml#L64
+
+### Risks and Mitigations
+
+#### Specific CCM doesn’t work properly on OpenShift
+
+The `CCMs` have not been tested on OCP yet. Unknown issues may arise.
+
+Severity: medium-high
+Likelihood: medium
+
+#### OpenShift components can't work with the CCM properly
+
+Since OpenShift imposes additional limitations compared to Kubernetes, some collisions are possible.
+So far there are no CCMs available in OpenShift, and there can be some problems with operability.
+
+Severity: medium-high
+Likelihood: low
+
+#### CSI driver is not available in OpenShift
+
+CCM doesn't support the in-tree `Volume` manager implementation. A CSI driver now becomes a mandatory requirement for every out-of-tree provider. Additionally the CSI should support data migration between the in-tree storage solution and the plugin provided storage interfaces.
+
+Severity: medium-high
+Likelihood: low
+
+### Test Plan
+
+- Make sure the `CCM` implementation for "any" cloud is up and running in a bootstrap phase before or shortly after the `kubelet` is running on any `Node`.
+- Ensure that transition between the in-tree and the out-of-tree will be handled by the established architecture with the new release payload, where all of the existing `Nodes` are operational during and after the upgrade time, and the bootstrap procedure for new ones is successful.
+- Ensure that upgrade from a release running on the in-tree provider only to the out-of-tree succeeds, and the downgrade is also supported.
+- Each provider should have a working CI implementation, which will assist in testing the provisioning and upgrades with out-of-tree support enabled.
+
+### Graduation Criteria
+
+#### Tech preview
+
+Operator handles OpenStack installation in 4.7.
+
+#### Tech Preview -> GA
+
+The operator will handle basic tasks, such as create initial templates at bootstrap and running all of supported in-tree providers after installation.
+
+## FAQ
+
+Q: Can we disaster-recover a cluster without CCM running?
+A: We can't. Basically, without CCM running, nodes can only join the cluster, but they will be unschedulable. Which means it's impossible to start any workloads on the nodes. [Source](https://kubernetes.io/docs/tasks/administer-cluster/running-cloud-controller/#running-cloud-controller-manager)
+
+Q: What is non-functional in a cluster during bootstrapping until CCM is up?
+A: If CCM is not available, new nodes in the cluster will be left unschedulable.
+
+Q: Who talks to CCM?
+A: From an architectural point of view, only Kube API server is able to communicate with CCM. [Source](https://kubernetes.io/docs/concepts/architecture/cloud-controller/#design). All requests to CCM are sent by `kubelet`. KCM doesn't interact with CCM.
```

Comment:
That picture is wrong, ccm talks to kas, to update taints on node, for starters. What about kubelet and kcm?

Reply:
Please open an upstream PR fixing this picture after you identify the correct order of the arrows on that picture.

Reply:
Agreed this is wrong and needs an update.

In the upstream picture they have it as if KA talks to KCM and KS, which is also wrong, the arrow from KA to CCM is consistent with this though